### PR TITLE
Avoid query type in generics

### PR DESCRIPTION
**Pull request by cjgillot**
There are at the moment roughly 170 queries in librustc.
The way ty::query is structured, a lot of code is duplicated for each query.
I suspect this to be responsible for a part of librustc'c compile time.

This PR reduces the amount of code generic on the query,
replacing it by code generic on the key-value types.

This is split out of #69808,
and should not contain the perf regression.

cc #65031

### CODE REVIEW

File: src/librustc/ty/query/plumbing.rs [hunk 1]
```diff
@@ -94,12 +97,42 @@ impl<'tcx, M: QueryAccessors<'tcx>> QueryState<'tcx, M> {
         let shards = self.shards.lock_shards();
         shards.iter().all(|shard| shard.active.is_empty())
     }
+
+    #[inline]
```
Comment (Zoxc) [inline src/librustc/ty/query/plumbing.rs]: This is cold, so `#[inline]` isn't needed.

**Comment by Zoxc** [comment]
I'm not sure how many unique key / value pairs we have for queries, if that isn't much lower than the total query count, this change isn't too impactful.

**Comment by cjgillot** [comment]
As a crude measurement, the number of instances of `alloc_self_profile_query_strings_for_query_cache` in the rlib goes from 201 to 148. `QueryState::iter_results` and `<DefaultCache as QueryCache>::iter` are called several times with closures, so go from 802 to 590 instances each.

**Comment by Zoxc** [comment]
That suggests that 26% of query key / value pairs are duplicates, so this is probably worthwhile.

File: src/librustc/ty/query/plumbing.rs [hunk 2]
```diff
@@ -955,9 +966,18 @@ macro_rules! define_queries_inner {
         // predictable symbol name prefix for query providers. This is helpful
         // for analyzing queries in profilers.
         pub(super) mod __query_compute {
```
Comment (Zoxc) [inline src/librustc/ty/query/plumbing.rs]: @mw @nnethercote @rust-lang/wg-compiler-performance Have this been used lately or could we just remove this and rely on `-Z self-profile`?

Comment (nnethercote) [reply src/librustc/ty/query/plumbing.rs]: I have never (knowingly) used this.

Comment (michaelwoerister) [reply src/librustc/ty/query/plumbing.rs]: I think @nikomatsakis perf-focus tool might have used this. I'm OK with removing it.

File: src/librustc/ty/query/plumbing.rs [hunk 3]
```diff
@@ -1006,9 +987,16 @@ macro_rules! define_queries_inner {
                 DepConstructor::$node(tcx, *key)
             }
 
-            #[inline]
+            #[inline(never)]
```
Comment (Zoxc) [inline src/librustc/ty/query/plumbing.rs]: This should be inlined.

**Comment by Zoxc** [comment]
Why is my `Erase` trait added here?

**Comment by cjgillot** [comment]
I was experimenting with it yesterday, and somehow it ended up on the wrong branch. Sorry.

File: src/librustc/ty/query/plumbing.rs [hunk 4]
```diff
@@ -94,12 +97,41 @@ impl<'tcx, M: QueryAccessors<'tcx>> QueryState<'tcx, M> {
         let shards = self.shards.lock_shards();
         shards.iter().all(|shard| shard.active.is_empty())
     }
+
+    pub(super) fn try_collect_active_jobs(
+        &self,
+        kind: DepKind,
+        make_query: impl Fn(C::Key) -> Query<'tcx> + Copy,
```
Comment (Zoxc) [inline src/librustc/ty/query/plumbing.rs]: You could use `&dyn` for `make_query` here.

**Comment by Zoxc** [comment]
> There are at the moment roughly 170 queries in librustc.
> The way ty::query is structured, a lot of code is duplicated for each query.
> I suspect this to be responsible for a part of librustc'c compile time.

I measured the compilation impact of the query system to be 25% of librustc's compile time 3. February.

File: src/librustc/ty/query/caches.rs [hunk 5]
```diff
@@ -23,7 +23,7 @@ pub(crate) trait QueryCache: Default {
     /// to compute it.
     fn lookup<'tcx, R, GetCache, OnHit, OnMiss>(
         &self,
-        state: &'tcx QueryState<'tcx, Self>,
+        lookup: QueryLookup<'tcx, Self::Key, Self::Sharded>,
```
Comment (Zoxc) [inline src/librustc/ty/query/caches.rs]: `QueryLookup` should not be passed here has `lookup` is designed to allow lock-free lookup of query caches while `QueryLookup` requires getting a lock. I'd just remove this commit.

**Comment by cjgillot** [comment]
Rebased.

**Comment by Zoxc** [comment]
I'd leave out the `Bundle key` commits since they add additional `clone` calls, otherwise this looks good.

**Comment by cjgillot** [comment]
Done.

File: src/librustc/ty/query/caches.rs [hunk 6]
```diff
@@ -1,82 +1,94 @@
...
+    type Value = V;
     type Sharded = FxHashMap<K, (V, DepNodeIndex)>;
 
     #[inline(always)]
-    fn lookup<'tcx, R, GetCache, OnHit, OnMiss, Q>(
+    fn lookup<'tcx, R, GetCache, OnHit, OnMiss>(
         &self,
-        state: &'tcx QueryState<'tcx, Q>,
+        state: &'tcx QueryState<'tcx, Self>,
         get_cache: GetCache,
```
Comment (Zoxc) [inline src/librustc/ty/query/caches.rs]: Now that you connected up the types, you should be able to just get rid of `get_cache: GetCache`.

File: src/librustc/ty/query/plumbing.rs [hunk 7]
```diff
@@ -20,44 +20,47 @@ use rustc_errors::{struct_span_err, Diagnostic, DiagnosticBuilder, FatalError, H
...
     #[cfg(debug_assertions)]
     pub(super) cache_hits: AtomicUsize,
 }
 
-impl<'tcx, Q: QueryAccessors<'tcx>> QueryState<'tcx, Q> {
-    pub(super) fn get_lookup<K: Hash>(&'tcx self, key: &K) -> QueryLookup<'tcx, Q> {
+impl<'tcx, C: QueryCache> QueryState<'tcx, C> {
+    pub(super) fn get_lookup<K2: Hash>(
+        &'tcx self,
+        key: &K2,
```
Comment (Zoxc) [inline src/librustc/ty/query/plumbing.rs]: `K2` can now be `C::Key`.

**Pull request merged**